pkg/operation/care/garden_health.go

```diff
@@ -0,0 +1,189 @@
+	newSystemComponentsCondition, err := h.checkGardenSystemComponents(ctx, checker, systemComponentsCondition)
+	newVirtualGardenComponentsCondition, err2 := h.checkVirtualGardenComponents(ctx, checker, virtualGardenComponentsCondition)
+
+	return []gardencorev1beta1.Condition{
+		NewConditionOrError(h.clock, systemComponentsCondition, newSystemComponentsCondition, err),
+		NewConditionOrError(h.clock, virtualGardenComponentsCondition, newVirtualGardenComponentsCondition, err2),
+	}
+}
+
+func (h *GardenHealth) checkGardenSystemComponents(
+	ctx context.Context,
+	checker *HealthChecker,
+	condition gardencorev1beta1.Condition,
+) (
+	*gardencorev1beta1.Condition,
+	error,
+) {
+	managedResources := sets.List(requiredGardenRuntimeManagedResources)
+	managedResources = append(managedResources, istio.ManagedResourceNames(true, "virtual-garden-")...)
+
+	if features.DefaultFeatureGate.Enabled(features.HVPA) {
+		managedResources = append(managedResources, hvpa.ManagedResourceName)
+	}
+	if h.garden.Spec.RuntimeCluster.Settings != nil &&
+		h.garden.Spec.RuntimeCluster.Settings.VerticalPodAutoscaler != nil &&
+		pointer.BoolDeref(h.garden.Spec.RuntimeCluster.Settings.VerticalPodAutoscaler.Enabled, false) {
+		managedResources = append(managedResources, vpa.ManagedResourceControlName)
+	}
+
+	for _, name := range managedResources {
+		namespace := h.gardenNamespace
+		if sets.New(istio.ManagedResourceNames(true, "virtual-garden-")...).Has(name) {
+			namespace = v1beta1constants.IstioSystemNamespace
+		}
+
+		mr := &resourcesv1alpha1.ManagedResource{}
+		if err := h.runtimeClient.Get(ctx, kubernetesutils.Key(namespace, name), mr); err != nil {
+			if apierrors.IsNotFound(err) {
+				exitCondition := checker.FailedCondition(condition, "ResourceNotFound", err.Error())
+				return &exitCondition, nil
+			}
+			return nil, err
+		}
+
+		if exitCondition := checkManagedResourceForGarden(checker, condition, mr); exitCondition != nil {
+			return exitCondition, nil
+		}
+	}
+
+	c := v1beta1helper.UpdatedConditionWithClock(h.clock, condition, gardencorev1beta1.ConditionTrue, "SystemComponentsRunning", "All system components are healthy.")
+	return &c, nil
+}
+
+func (h *GardenHealth) checkVirtualGardenComponents(
+	ctx context.Context,
+	checker *HealthChecker,
+	condition gardencorev1beta1.Condition,
+) (
+	*gardencorev1beta1.Condition,
+	error,
+) {
+	managedResources := sets.List(requiredVirtualGardenManagedResources)
+
+	for _, name := range managedResources {
+		namespace := h.gardenNamespace
+		mr := &resourcesv1alpha1.ManagedResource{}
+		if err := h.runtimeClient.Get(ctx, kubernetesutils.Key(namespace, name), mr); err != nil {
+			if apierrors.IsNotFound(err) {
+				exitCondition := checker.FailedCondition(condition, "ResourceNotFound", err.Error())
+				return &exitCondition, nil
+			}
+			return nil, err
+		}
+
+		if exitCondition := checkManagedResourceForGarden(checker, condition, mr); exitCondition != nil {
+			return exitCondition, nil
+		}
+	}
```

Comment:
Would it make sense to extract this common loop from both `checkGardenSystemComponents` and `checkVirtualGardenComponents`? The only special handling I see is with regards to `istio` and would work in both cases.